File: jetbrains/src/main/kotlin/deplens/utils/resolver/MavenRepoResolver.kt
package deplens.utils.resolver

import deplens.utils.service.RepoKey
import java.io.File

object MavenRepoResolver {

    fun repoKeyFromGroupArtifact(groupId: String?, artifactId: String?, version: String? = null): RepoKey? {
        if (groupId.isNullOrBlank() || artifactId.isNullOrBlank()) return null

        val resolvedVersion = resolveLocalVersion(groupId, artifactId, version) ?: return null
        return resolveRepoFromPom(groupId, artifactId, resolvedVersion, 0)
    }

    fun repoKeyFromResolvedPath(path: String?): RepoKey? {
        if (path.isNullOrBlank()) return null
        return repoKeyFromMavenLocalPath(path) ?: repoKeyFromGradleCachePath(path)
    }

    private fun repoKeyFromMavenLocalPath(path: String): RepoKey? {
        val marker = "/.m2/repository/"
        val idx = path.indexOf(marker)
        if (idx < 0) return null

        val after = path.substring(idx + marker.length)
        val jarPart = after.substringBefore('!')
        val parts = jarPart.split('/').filter { it.isNotBlank() }
        if (parts.size < 4) return null

        val artifactId = parts[parts.size - 3]
        val version = parts[parts.size - 2]
        val groupParts = parts.subList(0, parts.size - 3)
        val groupId = groupParts.joinToString(".")
        if (groupId.isBlank() || artifactId.isBlank() || version.isBlank()) return null

        return repoKeyFromGroupArtifact(groupId, artifactId, version)
    }

    private fun repoKeyFromGradleCachePath(path: String): RepoKey? {
        val marker = "/.gradle/caches/modules-2/files-2.1/"
        val idx = path.indexOf(marker)
        if (idx < 0) return null

        val after = path.substring(idx + marker.length)
        val jarPart = after.substringBefore('!')
        val parts = jarPart.split('/').filter { it.isNotBlank() }
        if (parts.size < 4) return null

        val groupId = parts[0]
        val artifactId = parts[1]
        val version = parts[2]
        val hash = parts[3]
        if (groupId.isBlank() || artifactId.isBlank() || version.isBlank()) return null

        val base = gradleCacheBase() ?: return null
        val dir = File(base, "$groupId/$artifactId/$version/$hash")
        val pom = File(dir, "$artifactId-$version.pom")

        val pomFile = when {
            pom.exists() -> pom
            else -> findPomInGradleCacheDir(base, groupId, artifactId, version)
        }

        if (pomFile != null && pomFile.exists()) {
            val content = runCatching { pomFile.readText() }.getOrNull() ?: return null
            val direct = extractGithubRepoKey(content)
            if (direct != null) return direct

            val parent = extractParentCoords(content) ?: return null
            if (parent.groupId.contains('$') || parent.artifactId.contains('$') || parent.version.contains('$')) return null

            return resolveRepoFromPom(parent.groupId, parent.artifactId, parent.version, 0)
        }

        val jar = File(dir, "$artifactId-$version.jar")
        if (jar.exists()) {
            val fromJar = extractGithubRepoKeyFromJar(jar)
            if (fromJar != null) return fromJar
        }

        return repoKeyFromGroupArtifact(groupId, artifactId, version)
    }

    private fun resolveLocalVersion(groupId: String, artifactId: String, version: String?): String? {
        val cleaned = version?.trim()?.takeIf { it.isNotBlank() && !it.contains('$') }
        if (cleaned != null) return cleaned

        val metaFile = localMetadataFile(groupId, artifactId) ?: return null
        if (!metaFile.exists()) return null

        val meta = runCatching { metaFile.readText() }.getOrNull() ?: return null

        val release = Regex("<release>([^<]+)</release>").find(meta)?.groupValues?.get(1)?.trim()
        if (!release.isNullOrBlank()) return release

        val latest = Regex("<latest>([^<]+)</latest>").find(meta)?.groupValues?.get(1)?.trim()
        if (!latest.isNullOrBlank()) return latest

        val versions = Regex("<version>([^<]+)</version>")
            .findAll(meta)
            .mapNotNull { it.groupValues.getOrNull(1)?.trim() }
            .filter { it.isNotBlank() }
            .toList()
        return versions.lastOrNull()
    }

    private fun localPomFile(groupId: String, artifactId: String, version: String): File? {
        val base = localRepoBase() ?: return null
        val groupPath = groupId.replace('.', File.separatorChar)
        return File(base, "$groupPath/$artifactId/$version/$artifactId-$version.pom")
    }

    private fun localMetadataFile(groupId: String, artifactId: String): File? {
        val base = localRepoBase() ?: return null
        val groupPath = groupId.replace('.', File.separatorChar)
        return File(base, "$groupPath/$artifactId/maven-metadata-local.xml")
            .takeIf { it.exists() }
            ?: File(base, "$groupPath/$artifactId/maven-metadata.xml")
    }

    private fun localRepoBase(): File? {
        val home = System.getProperty("user.home") ?: return null
        return File(home, ".m2/repository")
    }

    private fun gradleCacheBase(): File? {
        val home = System.getProperty("user.home") ?: return null
        return File(home, ".gradle/caches/modules-2/files-2.1")
    }

    private fun findPomInGradleCacheDir(
        base: File,
        groupId: String,
        artifactId: String,
        version: String
    ): File? {
        val versionDir = File(base, "$groupId/$artifactId/$version")
        val hashDirs = versionDir.listFiles()?.filter { it.isDirectory } ?: return null
        for (hashDir in hashDirs) {
            val pom = File(hashDir, "$artifactId-$version.pom")
            if (pom.exists()) return pom
        }
        return null
    }

    private fun extractGithubRepoKeyFromJar(jar: File): RepoKey? {
        return runCatching {
            java.util.zip.ZipFile(jar).use { zip ->
                val entries = zip.entries().asSequence().toList()
                val pomXmlEntry = entries.firstOrNull {
                    it.name.startsWith("META-INF/maven/") && it.name.endsWith("/pom.xml")
                }
                if (pomXmlEntry != null) {
                    val text = zip.getInputStream(pomXmlEntry).bufferedReader().use { it.readText() }
                    extractGithubRepoKey(text)?.let { return it }
                    val parent = extractParentCoords(text)
                    if (parent != null &&
                        !parent.groupId.contains('$') &&
                        !parent.artifactId.contains('$') &&
                        !parent.version.contains('$')
                    ) {
                        return resolveRepoFromPom(parent.groupId, parent.artifactId, parent.version, 0)
                    }
                }
                null
            }
        }.getOrNull()
    }

    private fun resolveRepoFromPom(groupId: String, artifactId: String, version: String, depth: Int): RepoKey? {
        if (depth > 6) return null

        val pom = localPomFile(groupId, artifactId, version) ?: return null
        if (!pom.exists()) return null

        val content = runCatching { pom.readText() }.getOrNull() ?: return null
        val direct = extractGithubRepoKey(content)
        if (direct != null) return direct

        val parent = extractParentCoords(content) ?: return null
        if (parent.groupId.contains('$') || parent.artifactId.contains('$') || parent.version.contains('$')) return null

        return resolveRepoFromPom(parent.groupId, parent.artifactId, parent.version, depth + 1)
    }

    private fun extractGithubRepoKey(text: String): RepoKey? {
        val scmBlock = Regex("<scm>.*?</scm>", RegexOption.DOT_MATCHES_ALL)
            .find(text)
            ?.value

        if (scmBlock != null) {
            val scmMatch = extractGithubRepoKeyFromText(scmBlock)
            if (scmMatch != null) return scmMatch
        }

        val url = Regex("<url>([^<]+)</url>").find(text)?.groupValues?.get(1)?.trim()
        if (!url.isNullOrBlank()) {
            val urlMatch = extractGithubRepoKeyFromText(url)
            if (urlMatch != null) return urlMatch
        }

        return extractGithubRepoKeyFromText(text)
    }

    private fun extractGithubRepoKeyFromText(text: String): RepoKey? {
        val match = Regex("github\\.com[:/]+([A-Za-z0-9_.-]+)/([A-Za-z0-9_.-]+)")
            .find(text) ?: return null

        val owner = match.groupValues[1]
        val repo = match.groupValues[2].removeSuffix(".git")
        if (owner.isBlank() || repo.isBlank()) return null
        return RepoKey(owner, repo)
    }

    private data class ParentCoords(val groupId: String, val artifactId: String, val version: String)

    private fun extractParentCoords(text: String): ParentCoords? {
        val parentBlock = Regex("<parent>.*?</parent>", RegexOption.DOT_MATCHES_ALL)
            .find(text)
            ?.value
            ?: return null

        val groupId = Regex("<groupId>([^<]+)</groupId>").find(parentBlock)?.groupValues?.get(1)?.trim()
        val artifactId = Regex("<artifactId>([^<]+)</artifactId>").find(parentBlock)?.groupValues?.get(1)?.trim()
        val version = Regex("<version>([^<]+)</version>").find(parentBlock)?.groupValues?.get(1)?.trim()

        if (groupId.isNullOrBlank() || artifactId.isNullOrBlank() || version.isNullOrBlank()) return null
        return ParentCoords(groupId, artifactId, version)
    }
}

File: vscode/src/utils/resolver/MavenRepoResolver.ts
import * as fs from "node:fs/promises";
import * as os from "node:os";
import * as path from "node:path";
import { RepoKey } from "../service/RepoKey";

export class MavenRepoResolver {
  static async repoKeyFromGroupArtifact(
    groupId?: string | null,
    artifactId?: string | null,
    version?: string | null,
  ): Promise<RepoKey | null> {
    if (!groupId || !artifactId) {
      return null;
    }

    const resolvedVersion = await this.resolveLocalVersion(groupId, artifactId, version);
    if (!resolvedVersion) {
      return null;
    }

    return this.resolveRepoFromPom(groupId, artifactId, resolvedVersion, 0);
  }

  static async repoKeyFromResolvedPath(filePath?: string | null): Promise<RepoKey | null> {
    if (!filePath) {
      return null;
    }

    return (
      (await this.repoKeyFromMavenLocalPath(filePath)) ??
      (await this.repoKeyFromGradleCachePath(filePath))
    );
  }

  private static async repoKeyFromMavenLocalPath(filePath: string): Promise<RepoKey | null> {
    const marker = `${path.sep}.m2${path.sep}repository${path.sep}`;
    const markerIndex = filePath.indexOf(marker);
    if (markerIndex < 0) {
      return null;
    }

    const normalizedPath = filePath.replace(/\\/g, "/");
    const normalizedMarker = "/.m2/repository/";
    const normalizedMarkerIndex = normalizedPath.indexOf(normalizedMarker);
    if (normalizedMarkerIndex < 0) {
      return null;
    }

    const after = normalizedPath.slice(normalizedMarkerIndex + normalizedMarker.length).split("!")[0];
    const parts = after.split("/").filter(Boolean);
    if (parts.length < 4) {
      return null;
    }

    const artifactId = parts[parts.length - 3];
    const version = parts[parts.length - 2];
    const groupId = parts.slice(0, parts.length - 3).join(".");

    return this.repoKeyFromGroupArtifact(groupId, artifactId, version);
  }

  private static async repoKeyFromGradleCachePath(filePath: string): Promise<RepoKey | null> {
    const normalizedPath = filePath.replace(/\\/g, "/");
    const marker = "/.gradle/caches/modules-2/files-2.1/";
    const markerIndex = normalizedPath.indexOf(marker);
    if (markerIndex < 0) {
      return null;
    }

    const after = normalizedPath.slice(markerIndex + marker.length).split("!")[0];
    const parts = after.split("/").filter(Boolean);
    if (parts.length < 4) {
      return null;
    }

    const [groupId, artifactId, version, hash] = parts;
    const base = this.gradleCacheBase();
    const candidatePom = path.join(base, groupId, artifactId, version, hash, `${artifactId}-${version}.pom`);
    const pomFile = (await this.pathExists(candidatePom))
      ? candidatePom
      : await this.findPomInGradleCacheDir(base, groupId, artifactId, version);

    if (pomFile) {
      const content = await this.readText(pomFile);
      if (content) {
        const direct = this.extractGithubRepoKey(content);
        if (direct) {
          return direct;
        }

        const parent = this.extractParentCoords(content);
        if (parent && !this.hasTemplateVar(parent.groupId, parent.artifactId, parent.version)) {
          return this.resolveRepoFromPom(parent.groupId, parent.artifactId, parent.version, 0);
        }
      }
    }

    return this.repoKeyFromGroupArtifact(groupId, artifactId, version);
  }

  private static async resolveLocalVersion(
    groupId: string,
    artifactId: string,
    version?: string | null,
  ): Promise<string | null> {
    const cleaned = version?.trim();
    if (cleaned && !cleaned.includes("$")) {
      return cleaned;
    }

    const metadataFile = await this.localMetadataFile(groupId, artifactId);
    if (!metadataFile) {
      return null;
    }

    const metadata = await this.readText(metadataFile);
    if (!metadata) {
      return null;
    }

    const release = metadata.match(/<release>([^<]+)<\/release>/)?.[1]?.trim();
    if (release) {
      return release;
    }

    const latest = metadata.match(/<latest>([^<]+)<\/latest>/)?.[1]?.trim();
    if (latest) {
      return latest;
    }

    const versions = [...metadata.matchAll(/<version>([^<]+)<\/version>/g)]
      .map((match) => match[1]?.trim())
      .filter((value): value is string => Boolean(value));

    return versions.at(-1) ?? null;
  }

  private static async resolveRepoFromPom(
    groupId: string,
    artifactId: string,
    version: string,
    depth: number,
  ): Promise<RepoKey | null> {
    if (depth > 6) {
      return null;
    }

    const pomFile = this.localPomFile(groupId, artifactId, version);
    if (!(await this.pathExists(pomFile))) {
      return null;
    }

    const content = await this.readText(pomFile);
    if (!content) {
      return null;
    }

    const direct = this.extractGithubRepoKey(content);
    if (direct) {
      return direct;
    }

    const parent = this.extractParentCoords(content);
    if (!parent || this.hasTemplateVar(parent.groupId, parent.artifactId, parent.version)) {
      return null;
    }

    return this.resolveRepoFromPom(parent.groupId, parent.artifactId, parent.version, depth + 1);
  }

  private static extractGithubRepoKey(text: string): RepoKey | null {
    const scmBlock = text.match(/<scm>[\s\S]*?<\/scm>/)?.[0];
    if (scmBlock) {
      const fromScm = this.extractGithubRepoKeyFromText(scmBlock);
      if (fromScm) {
        return fromScm;
      }
    }

    const url = text.match(/<url>([^<]+)<\/url>/)?.[1]?.trim();
    if (url) {
      const fromUrl = this.extractGithubRepoKeyFromText(url);
      if (fromUrl) {
        return fromUrl;
      }
    }

    return this.extractGithubRepoKeyFromText(text);
  }

  private static extractGithubRepoKeyFromText(text: string): RepoKey | null {
    const match = text.match(/github\.com[:/]+([A-Za-z0-9_.-]+)\/([A-Za-z0-9_.-]+)/);
    if (!match) {
      return null;
    }

    const owner = match[1];
    const repo = match[2]?.replace(/\.git$/, "");
    if (!owner || !repo) {
      return null;
    }

    return { owner, repo };
  }

  private static extractParentCoords(text: string): { groupId: string; artifactId: string; version: string } | null {
    const parentBlock = text.match(/<parent>[\s\S]*?<\/parent>/)?.[0];
    if (!parentBlock) {
      return null;
    }

    const groupId = parentBlock.match(/<groupId>([^<]+)<\/groupId>/)?.[1]?.trim();
    const artifactId = parentBlock.match(/<artifactId>([^<]+)<\/artifactId>/)?.[1]?.trim();
    const version = parentBlock.match(/<version>([^<]+)<\/version>/)?.[1]?.trim();

    if (!groupId || !artifactId || !version) {
      return null;
    }

    return { groupId, artifactId, version };
  }

  private static localPomFile(groupId: string, artifactId: string, version: string): string {
    return path.join(
      this.localRepoBase(),
      groupId.replace(/\./g, path.sep),
      artifactId,
      version,
      `${artifactId}-${version}.pom`,
    );
  }

  private static async localMetadataFile(groupId: string, artifactId: string): Promise<string | null> {
    const artifactDir = path.join(this.localRepoBase(), groupId.replace(/\./g, path.sep), artifactId);
    const localMetadata = path.join(artifactDir, "maven-metadata-local.xml");
    if (await this.pathExists(localMetadata)) {
      return localMetadata;
    }

    const metadata = path.join(artifactDir, "maven-metadata.xml");
    return (await this.pathExists(metadata)) ? metadata : null;
  }

  private static localRepoBase(): string {
    return path.join(os.homedir(), ".m2", "repository");
  }

  private static gradleCacheBase(): string {
    return path.join(os.homedir(), ".gradle", "caches", "modules-2", "files-2.1");
  }

  private static async findPomInGradleCacheDir(
    base: string,
    groupId: string,
    artifactId: string,
    version: string,
  ): Promise<string | null> {
    const versionDir = path.join(base, groupId, artifactId, version);
    try {
      const entries = await fs.readdir(versionDir, { withFileTypes: true });
      for (const entry of entries) {
        if (!entry.isDirectory()) {
          continue;
        }

        const pomFile = path.join(versionDir, entry.name, `${artifactId}-${version}.pom`);
        if (await this.pathExists(pomFile)) {
          return pomFile;
        }
      }
    } catch {
      return null;
    }

    return null;
  }

  private static async readText(filePath: string): Promise<string | null> {
    try {
      return await fs.readFile(filePath, "utf8");
    } catch {
      return null;
    }
  }

  private static async pathExists(filePath: string): Promise<boolean> {
    try {
      await fs.access(filePath);
      return true;
    } catch {
      return false;
    }
  }

  private static hasTemplateVar(...values: string[]): boolean {
    return values.some((value) => value.includes("$"));
  }
}
